onnxruntime/core/providers/nnapi/nnapi_builtin/builders/model_builder.cc

```diff
@@ -118,74 +118,89 @@ Status ModelBuilder::GetTargetDevices() {
         continue;
 
       if (min == -1.0f && max == 1.0f) {
-        activation_nodes_.emplace(node->Index(), ANEURALNETWORKS_FUSED_RELU1);
+        activation_node_units_.emplace(node_unit.get(), ANEURALNETWORKS_FUSED_RELU1);
       } else if (min == 0.0f && max == 6.0f) {
-        activation_nodes_.emplace(node->Index(), ANEURALNETWORKS_FUSED_RELU6);
+        activation_node_units_.emplace(node_unit.get(), ANEURALNETWORKS_FUSED_RELU6);
       }
     }
   }
 }
 
-// Help to get all quantized operators' input and the node(s) using the input
-static std::unordered_map<std::string, vector<const Node*>> GetAllQuantizedOpInputs(const GraphViewer& graph_viewer) {
-  std::unordered_map<std::string, vector<const Node*>> all_quantized_op_inputs;
-  const auto& node_indices = graph_viewer.GetNodesInTopologicalOrder();
-  for (const auto& node_idx : node_indices) {
-    const auto* node(graph_viewer.GetNode(node_idx));
-    auto qlinear_op_type = GetQLinearOpType(*node);
+const NodeUnit& ModelBuilder::GetNodeUnit(const Node* node) const {
+  // In theory, if node_unit_map_ is generated correctly, see PreprocessNodeUnits(), a NodeUnit can be
+  // found for any single node in the graph_viewer_, unless the given node is not from graph_viewer_
+  return *node_unit_map_.at(node);
+}
+
+void ModelBuilder::PreprocessNodeUnits() {
+  // TODO, hookup shared QDQ selectors here to identify all the qdq NodeUnit in the graph
+  const auto& node_indices = graph_viewer_.GetNodesInTopologicalOrder();
+  for (size_t i = 0; i < node_indices.size(); i++) {
+    const auto node_idx = node_indices[i];
+    // TODO, check if the node is already part of a qdq group
+    const auto* node(graph_viewer_.GetNode(node_idx));
+    auto node_unit = std::make_unique<NodeUnit>(*node);
+    node_unit_map_.insert({node, node_unit.get()});
+    node_unit_holder_.push_back(std::move(node_unit));
+  }
+}
+
+// Help to get all quantized operators' input and the NodeUnit(s) using the input
+void ModelBuilder::GetAllQuantizedOpInputs() {
+  for (const auto& node_unit : node_unit_holder_) {
+    // TODO, hookup getting quantized inputs with QDQ NodeUnits and remove the ORT_ENFORCE
+    ORT_ENFORCE(node_unit->UnitType() == NodeUnit::Type::SingleNode, "QDQ NodeUnit is not yet implemented");
+
+    auto qlinear_op_type = GetQLinearOpType(node_unit->GetNode());
 
     // Not a qlinear op
     if (qlinear_op_type == QLinearOpType::Unknown)
       continue;
 
+    const auto add_quantized_input =
+        [&all_quantized_op_inputs = all_quantized_op_inputs_](const NodeUnit& node_unit, size_t input_idx) {
+          const auto& input_name = node_unit.Inputs()[input_idx].node_arg.Name();
+          if (Contains(all_quantized_op_inputs, input_name))
+            all_quantized_op_inputs.at(input_name).push_back(&node_unit);
+          else
+            all_quantized_op_inputs.emplace(input_name, std::vector<const NodeUnit*>{&node_unit});
```

Comment:
nit: how about `all_quantized_op_inputs[input_name].push_back(&node_unit)` instead of `if (Contains(...)) ... else ...`